external/urlfrontier/src/main/java/com/digitalpebble/stormcrawler/urlfrontier/StatusUpdaterBolt.java

```diff
@@ -202,114 +262,170 @@ public void onCompleted() {
 
     @Override
     public void store(
-            String url, Status status, Metadata metadata, Optional<Date> nextFetch, Tuple t)
-            throws Exception {
-
-        int counter = 0;
-
-        while (messagesinFlight.get() >= this.maxMessagesinFlight) {
-            counter++;
-            LOG.debug(
-                    "{} messages in flight - waiting for {} msec",
-                    messagesinFlight.get(),
-                    throttleTime * counter);
-            eventCounter.scope("timeSpentThrottling").incrBy(throttleTime * counter);
-            Utils.sleep(throttleTime * counter);
+            @NotNull String url,
+            @NotNull Status status,
+            @NotNull Metadata metadata,
+            @NotNull Optional<Date> nextFetch,
+            @NotNull Tuple t) {
+
+        var hasPermit = false;
+        var timeSpent = 0L;
+
+        // First get processing permit. Otherwise, starvation possible.
+        while (!hasPermit) {
+            try {
+                hasPermit = inFlightSemaphore.tryAcquire(throttleTimeMS, TimeUnit.MILLISECONDS);
+                if (!hasPermit) {
+                    LOG.debug(
+                            "{} messages in flight, time spent throttling {}",
+                            inFlightSemaphore.getQueueLength(),
+                            timeSpent);
+                    eventCounter.scope("timeSpentThrottling").incrBy(throttleTimeMS);
+                    timeSpent += throttleTimeMS;
+                    if (timeSpent >= 30000L) {
+                        LOG.warn(
+                                "Waiting more than {} ms for processing. There are {} permits available for {} waiting threads.",
+                                timeSpent,
+                                inFlightSemaphore.availablePermits(),
+                                inFlightSemaphore.getQueueLength());
+                    }
+                }
+            } catch (InterruptedException e) {
+                LOG.info(
```

Comment:
shouldn't the interrupt status be restored?

`        Thread.currentThread().interrupt();
`

Reply:
> @FelixEngl do you have a Twitter account? I d like to mention your work there when it is merged

Yes, I actually have one :D
https://twitter.com/EnglFelix

Reply:
You are right, we have to handle an interrupt correctly. 

I'll look up what ApacheStorm wants in this case.

Depending on that interrupt should be enough, otherwise we have to clean up manually before either throwing a new interrupt or whatever.

Reply:
Interrupt is enough, thank you!